core/src/test/scala/integration/kafka/api/GroupAuthorizerIntegrationTest.scala

```diff
@@ -12,30 +12,139 @@
+    properties.put(KafkaConfig.AuthorizerClassNameProp, classOf[SimpleAclAuthorizer].getName)
+    properties.put(KafkaConfig.BrokerIdProp, brokerId.toString)
+    properties.put(KafkaConfig.OffsetsTopicPartitionsProp, "1")
+    properties.put(KafkaConfig.OffsetsTopicReplicationFactorProp, "1")
+    properties.put(KafkaConfig.TransactionsTopicPartitionsProp, "1")
+    properties.put(KafkaConfig.TransactionsTopicReplicationFactorProp, "1")
+    properties.put(KafkaConfig.TransactionsTopicMinISRProp, "1")
+    properties.put(BrokerSecurityConfigs.PRINCIPAL_BUILDER_CLASS_CONFIG, classOf[GroupPrincipalBuilder].getName)
+  }
+
+  @Before
+  override def setUp(): Unit = {
+    doSetup(createOffsetsTopic = false)
+
+    // Allow inter-broker communication
+    TestUtils.addAndVerifyAcls(servers.head,
+      Set(createAcl(AclOperation.CLUSTER_ACTION, AclPermissionType.ALLOW, principal = BrokerPrincipal)),
+      new ResourcePattern(ResourceType.CLUSTER, Resource.CLUSTER_NAME, PatternType.LITERAL))
+
+    TestUtils.createOffsetsTopic(zkClient, servers)
+  }
+
+  private def createAcl(aclOperation: AclOperation,
+                        aclPermissionType: AclPermissionType,
+                        principal: KafkaPrincipal = ClientPrincipal): AccessControlEntry = {
+    new AccessControlEntry(principal.toString, WildcardHost, aclOperation, aclPermissionType)
+  }
+
+  @Test
+  def testUnauthorizedProduceAndConsume(): Unit = {
+    val topic = "topic"
+    val topicPartition = new TopicPartition("topic", 0)
+
+    createTopic(topic)
+
+    val producer = createProducer()
+    intercept[TopicAuthorizationException] {
+      sendRecords(producer, numRecords = 10, topicPartition)
+    }
+
+    val consumer = createConsumer(configsToRemove = List(ConsumerConfig.GROUP_ID_CONFIG))
+    consumer.assign(List(topicPartition).asJava)
+    val e = intercept[TopicAuthorizationException] {
+      TestUtils.pollUntilAtLeastNumRecords(consumer, numRecords = 10)
+    }
+    assertEquals(Collections.singleton(topic), e.unauthorizedTopics())
+  }
+
+  @Test
+  def testAuthorizedProduceAndConsume(): Unit = {
+    val topic = "topic"
+    val topicPartition = new TopicPartition("topic", 0)
+
+    createTopic(topic)
+
+    TestUtils.addAndVerifyAcls(servers.head,
+      Set(createAcl(AclOperation.WRITE, AclPermissionType.ALLOW)),
+      new ResourcePattern(ResourceType.TOPIC, topic, PatternType.LITERAL))
+    val producer = createProducer()
+    sendRecords(producer, numRecords = 10, topicPartition)
+
+    TestUtils.addAndVerifyAcls(servers.head,
+      Set(createAcl(AclOperation.READ, AclPermissionType.ALLOW)),
+      new ResourcePattern(ResourceType.TOPIC, topic, PatternType.LITERAL))
+    val consumer = createConsumer(configsToRemove = List(ConsumerConfig.GROUP_ID_CONFIG))
+    consumer.assign(List(topicPartition).asJava)
+    TestUtils.pollUntilAtLeastNumRecords(consumer, numRecords = 10)
   }
+
+  private def sendRecords(producer: KafkaProducer[Array[Byte], Array[Byte]],
```

Comment:
We can't use one of the utility methods in `TestUtils`?

Reply:
Believe it or not, this patch actually preceded those methods 🙂